File: xju/io/test-FileReader.cc
// Copyright (c) 2017 Trevor Taylor
//
// Permission to use, copy, modify, distribute and sell this software
// and its documentation for any purpose is hereby granted without fee,
// provided that the above copyright notice appear in all.
// Trevor Taylor makes no representations about the suitability of this
// software for any purpose.  It is provided "as is" without express or
// implied warranty.
//
#include <xju/io/FileReader.hh>

#include <iostream>
#include <xju/assert.hh>
#include <xju/now.hh>
#include <xju/file/write.hh>
#include <xju/startsWith.hh>
#include <xju/file/touch.hh>
#include <xju/file/rm.hh>

namespace xju
{
namespace io
{

void test1() {
  auto const fileName(xju::path::split("test.txt"));
  xju::file::touch(fileName,xju::file::Mode(0777));
  xju::file::rm(fileName);
  try{
    FileReader x(fileName);
  }
  catch(xju::SyscallFailed const& e){
    xju::assert_equal(e._errno,ENOENT);
  }
  
  xju::file::write(
    fileName,
    "fred",4,
    xju::file::Mode(0777));

  FileReader x(fileName);
  std::string data(4U,' ');
  data.resize(x.read(&data[0],2));
  xju::assert_equal(data,"fr");

  data.resize(4U);
  data.resize(x.read(&data[0],3));
  xju::assert_equal(data,"ed");

  try{
    data.resize(1U);
    data.resize(x.read(&data[0],1));
    xju::assert_never_reached();
  }
  catch(xju::io::EndOfFile const& e){
    std::string const r(readableRepr(e));
    xju::assert_equal(xju::startsWith(r,std::string("Failed to read up to 1 bytes at offset 4 of file")),true);
  }

  x.seekBy(-1);
  data.resize(4U);
  data.resize(x.read(&data[0],1));
  xju::assert_equal(data,"d");
  
  x.seekTo(1);
  data.resize(4U);
  data.resize(x.read(&data[0],2));
  xju::assert_equal(data,"re");
  
}

}
}

using namespace xju::io;

int main(int argc, char* argv[])
{
  unsigned int n(0);
  test1(), ++n;
  std::cout << "PASS - " << n << " steps" << std::endl;
  return 0;
}

File: xju/cmc/io/test-FileReader.py
#!/usr/bin/env python3

# Copyright (c) 2022 Trevor Taylor
# coding: utf-8
# 
# Permission to use, copy, modify, and/or distribute this software for
# any purpose with or without fee is hereby granted, provided that all
# copyright notices and this permission notice appear in all copies.
# 
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

from xju.cmc.io import FileReader
from pathlib import Path
from xju.assert_ import Assert
from xju.xn import readableRepr

try:
    with FileReader(Path("xxx.txt")) as f:
        pass
except Exception as e:
    Assert(readableRepr(e))=="Failed to open xxx.txt reader with close-on-exec True because\n[Errno 2] No such file or directory: 'xxx.txt'."
else:
    assert False, f'should not be here with {f}'
    pass

with open('xxx.txt','w') as f2:
    f2.write('fredward')
    pass

with FileReader(Path("xxx.txt")) as f:
    Assert(f.input.readall().decode('utf-8'))=="fredward"
    Assert(f.size())==8
    Assert(f.seek_by(-2).input.read(2))==b'rd'
    Assert(f.seek_to(2).input.read(3))==b'edw'
    # seeking past end is not an error, even if it makes no sense
    Assert(f.seek_to(9).input.tell())==9

    try:
        f.seek_to(-1)
    except Exception as e:
        Assert(readableRepr(e).replace('\n',' ')).matches(
            'Failed to position so next write occurs -1 bytes from start of file because.* Invalid argument.')
    else:
        assert False, 'should not b here'
    pass

assert not hasattr(f,'input')
